### DISCUSSION TITLE
Slash encoding in qualifiers

### DISCUSSION THREAD

**Discussion by aloubyansky** [package-url/purl-spec#293] [PURL tests]
Continuing the discussion that started in https://github.com/package-url/packageurl-python/issues/121#issuecomment-1966641190

It appears different implementations under https://github.com/package-url implement encoding of slashes (among other things) differently. It'd be good to clarify whether some of them should be fixed to have a consistent encoding across different implementations.

For example, the Python implementation
```
from packageurl import PackageURL
purl = PackageURL.from_string("pkg:maven/org.apache.james/apache-mime4j-storage@0.8.9.redhat-00001?repository_url=https%3A%2F%2Fmaven.repository.redhat.com%2Fga%2F&type=jar")
print(purl.to_string())
```
results in
```
pkg:maven/org.apache.james/apache-mime4j-storage@0.8.9.redhat-00001?repository_url=https://maven.repository.redhat.com/ga/&type=jar
```
While the Java one
```
//DEPS com.github.package-url:packageurl-java:1.5.0

import java.util.TreeMap;
import com.github.packageurl.PackageURL;

public class purl {
    public static void main(String[] args) throws Exception {

        final TreeMap<String, String> qualifiers = new TreeMap<>();
        qualifiers.put("type", "jar");
        qualifiers.put("repository_url", "https://maven.repository.redhat.com/ga/");
        var purl = new PackageURL(PackageURL.StandardTypes.MAVEN,
                    "org.apache.james",
                    "apache-mime4j-storage",
                    "0.8.9.redhat-00001",
                    qualifiers, null);
        System.out.println(purl);
    }
}
```
results in
```
pkg:maven/org.apache.james/apache-mime4j-storage@0.8.9.redhat-00001?repository_url=https%3A%2F%2Fmaven.repository.redhat.com%2Fga%2F&type=jar
```
Would it be worth it creating an issue to align implementations of the encoding rules? Thanks.

FYI @tichavskym

**Comment by matt-phylum** [comment]
The official test suite makes it clear that slashes are expected not to be escaped.

For example (ironically for Java packages, even though the Java implementation does escape them):

https://github.com/package-url/purl-spec/blob/346589846130317464b677bc4eab30bf5040183a/test-suite-data.json#L110-L121

But I have noticed the following reasons for implementations to fail the official test suite:
- They don't use the official test suite.
- They use an old copy of the test suite.
- They ["correct"](https://github.com/package-url/packageurl-java/blob/47351deccffb9394df42b913f59ffc00741af0b6/src/test/resources/test-suite-data.json#L110-L121) the test suite.
- They use the test suite, but they skip tests for unsupported package types. This makes sense because some of the tests are testing type-specific rules. If your implementation doesn't support mlflow name normalization, you're not going to pass the mlflow tests, so just skip them all. Unfortunately, the test suite doesn't do a very good job of intentionally covering the core aspects such as escaping rules, so implementations that skip package types might be skipping tests that are relevant for other package types.
- They don't implement the test suite correctly. Implementations need to verify that parsing the provided PURL produces the component values, formatting the component values produces the canonical PURL, and finally that parsing the canonical PURL produces the same component values.

I made https://github.com/phylum-dev/purl-survey to be able to compare implementations and it can run the same test suite consistently for all implementations. I entered GitHub issues for some egregious implementation errors, but there are still [serious](https://github.com/package-url/purl-spec/pull/261) [problems](https://github.com/package-url/purl-spec/pull/273) which are not covered by the test suite and can cause PURLs to change meanings.

**Reply by aloubyansky** [reply]
@stevespringett wdyt? Worth creating an issue to fix it in https://github.com/package-url/packageurl-java? Thanks!

**Reply by aloubyansky** [reply]
I created https://github.com/package-url/packageurl-java/issues/122

**Reply by stevespringett** [reply]
The Java version was using a really old test suite. However, updating it causes a lot of issues. It turns out that the official test suite now includes purl type specific tests, which is not suppose to be in there. 

The test suite was intended to be used for testing the purl spec itself, not the individual purl types. I recommend we separate out the purl type tests into their own test suite. This will help as we version the purl spec independent of the purl types themselves, which may end up having their own registry in the future.

So, in order to fix this one issue affecting the core spec, I now need to fix a boatload of other issues, all of which are purl type specific, if the Java library is going to use the updated test suite.
